### PR TITLE
Depreciates `kornia.testing`

### LINKED ISSUE #2162: Add tensor to gradcheck  into the base tester
Tensor to gradcheck should be also in the base tester

_Originally posted by @edgarriba in https://github.com/kornia/kornia/pull/2152#discussion_r1071806021_
 
apply to all tensors on the inputs of `self.gradcheck` automatically

### PR DESCRIPTION
**Pull request by johnnv1**
#### Changes
This patch (bigger than expected) depreciates the `kornia.testing` module. Planning to delete it in December, before the 0.8.0 release.

The `kornia.testing` has been used as a set of functionalities for kornia test suite. Having it as a module, it makes some changes unviable, because needs to depreciate things before removing/changing.

The functions that have been used internally I moved it to live in another module, If you (@edgarriba @shijianjian @ducha-aiki) think that other features that exist in `kornia.testing` should continue to be part of the public package, let me know.

All the functionalities used in our test suite have been moved to the `testing` directory. At `testing` should live all the code that we need just for tests, but aren't tests themselves (Like common functions to generate data or tests).

Within this migration (kornia.testing -> testing) I fixes #2162, and also updated a lot of tests to use the `BaseTester`. I also updated the `BaseTester` to **not require the annoying tests**. and now it is just a base class within some functionalities (as the assert_close and gradcheck with magics helpers)

Summarizing: 
- depreciated `kornia.testing`
- move tests utilities code to `testing` directory -- which allows us to make changes more abruptly, without worrying about it being a feature of the library
- Update all tests and code (yes, the testing API have been used on some internal code) to not use the `kornia.testing`
- update a lot of `gradtest` to not run within float32, since they are always casting it to float64
- enhancement the `BaseTester` by:
  - add more magics under `assert_close` and `gradcheck` -- now it auto-cast all tensors to fp64 within grad by default (#2162)
  - remove the abstract methods. Making it easier to prototype new features, without the need for multiple (often pointless) tests

Things to be opened as issues for us to do after this PR:
- Move code that lives in `tests/` which are not tests themselves, to live in `testing`.
- Migrate all test classes to use `BaseTester` and its functionalities.
- delete `kornia.testing` in December 2024

#### Type of change
- [x] 🧪 Tests Cases

### CODE REVIEW

**Comment by johnnv1** [comment]
i'll finish to fix the failing tests throughout the week

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -162,7 +162,7 @@ maintainable.
     It is also a good idea to group the tests for functionalities
 
     ```python
-    from kornia.testing import BaseTester
+    from testing.base import BaseTester
```
Comment (johnnv1) [inline CONTRIBUTING.md:166]: add some text about the testing module itself for the community

File: kornia/__init__.py [hunk 2]
```diff
@@ -9,8 +9,7 @@
 
 # NOTE: we are going to expose to top level very few things
 from kornia.constants import pi
-from kornia.testing import xla_is_available
```
Comment (johnnv1) [inline kornia/__init__.py:12]: should we plan to do some work on the kornia imports for 0.8.0 release? to try to reduce the number of circular import

File: pyproject.toml [hunk 3]
```diff
@@ -145,7 +145,7 @@ requires = [
     skip-magic-trailing-comma = false
 
   [tool.ruff.isort]
-    known-first-party = ["examples", "kornia", "test"]
+    known-first-party = ["examples", "kornia", "tests", "testing"]
```
Comment (johnnv1) [inline pyproject.toml:148]: add a rule to split from testing and kornia imports by a empty line

File: tests/augmentation/test_augmentation.py [hunk 4]
```diff
@@ -55,28 +42,27 @@
...
 from kornia.utils.helpers import _torch_inverse_cast
-
-# TODO same_on_batch tests?
+from testing.augmentation.datasets import DummyMPDataset
+from testing.base import BaseTester
+from testing.overwrite import default_with_one_parameter_changed
 
 
 @pytest.mark.usefixtures("device", "dtype")
 class CommonTests(BaseTester):
```
Comment (johnnv1) [inline tests/augmentation/test_augmentation.py:63]: to do after this PR, move it to `testing`

File: kornia/__init__.py (hunk 2 above)
Comment (edgarriba) [reply kornia/__init__.py:12]: i think yes -- which possibly means at at least internally not use top level module imports

File: testing/error.py [hunk 5]
```diff
@@ -0,0 +1,10 @@
+from __future__ import annotations
+
+import torch
+
+from kornia.core import Tensor
+
+
+def compute_patch_abs_error(x: Tensor, y: Tensor, h: int, w: int) -> Tensor:
```
[the remark this answers is not in the document]
Comment (johnnv1) [reply testing/error.py:8]: i don't see a specific application of it

**Review by edgarriba — APPROVED** [review]

**Pull request merged**